File: src/main/scala/com/gigahex/samples/OSAnalysis.scala
package com.gigahex.samples

import org.apache.spark.sql.SparkSession
import nl.basjes.parse.useragent.UserAgentAnalyzer
import org.apache.spark.sql._
import org.apache.spark.sql.functions.desc


object OSAnalysis {

  def getDeviceInfo(row: Row): (String, String, String, String) = {
    val uaa = UserAgentAnalyzer
      .newBuilder()
      .hideMatcherLoadStats()
      .withCache(10000)
      .build()
    val result = uaa.parse(row.getAs[String]("user_agent"))
    (row.getAs[String]("user_id"),
      result.getValue("DeviceName"),
      result.getValue("AgentName"),
      result.getValue("OperatingSystemNameVersionMajor"))
  }


  def main(args: Array[String]): Unit = {
    /**
     * Setup the spark session
     */
    val spark = SparkSession.builder()
      .appName("devices_usage")
      .master("local")
      .getOrCreate()
    import spark.implicits._

    val websiteLogs = spark.read.json("hdfs://0.0.0.0:9075/user/gigahex/logs_devices.json")
    websiteLogs.show()
    val total = websiteLogs.count()
    val withDevices = websiteLogs
      .map(row => getDeviceInfo(row)).toDF("user_id", "device", "browser", "os")
      .cache()
    withDevices.show()

    val stats = withDevices.groupBy("os").count()
      .map(row => (row.getAs[String]("os"),
        row.getAs[Long]("count"),
        (row.getAs[Long]("count") / total.toDouble) * 100))
      .toDF("Operating System", "Users", "% Users")
      .orderBy(desc("Users"))

    stats.show()
    spark.stop()


  }

}

File: src/main/py/os_analysis.py
from pyspark.sql import SparkSession
from pyspark.sql.types import Row
from pyspark.sql.functions import desc
from ua_parser import user_agent_parser

# Create the spark session

spark = SparkSession.builder.master("local") \
    .appName("count-os-usage") \
    .getOrCreate()

website_logs = spark.read.json("hdfs://0.0.0.0:9075/user/gigahex/logs_devices.json")
with_os = website_logs \
    .rdd \
    .map(lambda row: Row(
    user_id=row.asDict()["user_id"],
    os=user_agent_parser.ParseOS(row.asDict()["user_agent"])['family'],
)).toDF()

with_os.show()

total_users = website_logs.count()
stats = with_os.groupBy("os") \
    .count() \
    .rdd \
    .map(lambda row: Row(operating_system=row[0],
                         users=row[1],
                         percentage_users=(row[1] / total_users) * 100)).toDF().sort(desc("users"))
stats.show()

spark.stop()
